src/main/java/org/prebid/server/bidder/loyal/LoyalBidder.java

```diff
@@ -0,0 +1,135 @@
+import org.prebid.server.bidder.model.BidderCall;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.loyal.ExtImpLoyal;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.BidderUtil;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+public class LoyalBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpLoyal>> LOYAL_EXT_TYPE_REFERENCE =
+            new TypeReference<>() {
+            };
+
+    private static final String PLACEMENT_ID_MACRO = "{{PlacementId}}";
+    private static final String ENDPOINT_ID_MACRO = "{{EndpointId}}";
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public LoyalBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<BidderError> errors = new ArrayList<>();
+        final List<HttpRequest<BidRequest>> requests = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            try {
+                final ExtImpLoyal ext = parseImpExt(imp);
+                final HttpRequest<BidRequest> httpRequest = createHttpRequest(ext, request);
+                requests.add(httpRequest);
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        if (!errors.isEmpty()) {
+            return Result.withErrors(errors);
+        }
+
+        return Result.withValues(requests);
+    }
+
+    private ExtImpLoyal parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), LOYAL_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException("Missing bidder ext in impression with id: " + imp.getId());
+        }
+    }
+
+    private HttpRequest<BidRequest> createHttpRequest(ExtImpLoyal ext, BidRequest request) {
+        String url = endpointUrl;
+        url = StringUtils.isNotBlank(ext.getPlacementId())
+                ? url.replace(PLACEMENT_ID_MACRO, ext.getPlacementId()) : url.replace("param={{PlacementId}}&", "");
+        url = StringUtils.isNotBlank(ext.getEndpointId())
+                ? url.replace(ENDPOINT_ID_MACRO, ext.getEndpointId()) : url.replace("&param2={{EndpointId}}", "");
```

Comment:
this is never happened in the Go version, their endpoint url doesn't have macros at all